ocaml/typing/typecore.ml

```diff
@@ -6665,10 +6665,11 @@ and type_function
                   suffix_contains_gadt
               in
               let curry =
-                if is_final_val_param then
+                match fun_alloc_mode with
+                | None ->
+                  assert(is_final_val_param);
```

Comment:
Isn't this just replacing option.get exception with assert exception?

Reply:
Yes, but `Option.get` is an anti-pattern, to me. That is, we have `option` so that we don't use `Option.get`.

Reply:
I think the root anti-pattern here is that `is_final_val_param` and `fun_alloc_mode` have duplicated information. And this anti-pattern manifests by either `Option.get` or `assert` - neither is better. But `is_final_val_param` wins slightly for its suggestive name.

Reply:
Well, maybe we can just remove the `assert`? Then I think it looks fine.

Reply:
I agree about the root anti-pattern here. I looked into fixing that but saw no easy way. I would personally keep the `assert`, because it helps the reader understand what's going on. Another possibility is to revert to the original code here but add a comment saying that the `Option.get` is safe because `is_final_val_param` is false.

I don't actually feel that strongly here. I will Resolve this conversation. Happy to merge with whatever here, if you want to make a change.

Reply:
Thanks - I added comment and a symmetric `assert`.